tests/components/songpal/test_media_player.py

```diff
@@ -0,0 +1,267 @@
+
+    await hass.services.async_call(
+        songpal.DOMAIN,
+        SET_SOUND_SETTING,
+        {"entity_id": "all", "name": "name", "value": "value"},
+        blocking=True,
+    )
+    mocked_device.set_sound_settings.assert_called_once_with("name", "value")
+    mocked_device2.set_sound_settings.assert_called_once_with("name", "value")
+
+
+async def test_websocket_events(hass):
+    """Test websocket events."""
+    mocked_device = _create_mocked_device()
+    entry = MockConfigEntry(domain=songpal.DOMAIN, data=CONF_DATA)
+    entry.add_to_hass(hass)
+
+    with _patch_media_player_device(mocked_device):
+        await hass.config_entries.async_setup(entry.entry_id)
+        await hass.async_block_till_done()
+
+    mocked_device.listen_notifications.assert_called_once()
+    assert mocked_device.on_notification.call_count == 4
+
+    notification_callbacks = mocked_device.notification_callbacks
+
+    volume_change = MagicMock()
+    volume_change.mute = True
+    volume_change.volume = 20
+    await notification_callbacks[VolumeChange](volume_change)
+    attributes = _get_attributes(hass)
+    assert attributes["is_volume_muted"] is True
+    assert attributes["volume_level"] == 0.2
+
+    content_change = MagicMock()
+    content_change.is_input = False
+    content_change.uri = "uri1"
+    await notification_callbacks[ContentChange](content_change)
+    assert _get_attributes(hass)["source"] == "title2"
+    content_change.is_input = True
+    await notification_callbacks[ContentChange](content_change)
+    assert _get_attributes(hass)["source"] == "title1"
+
+    power_change = MagicMock()
+    power_change.status = False
+    await notification_callbacks[PowerChange](power_change)
+    assert hass.states.get(ENTITY_ID).state == STATE_OFF
+
+
+async def test_disconnected(hass, caplog):
+    """Test disconnected behavior."""
+    mocked_device = _create_mocked_device()
+    entry = MockConfigEntry(domain=songpal.DOMAIN, data=CONF_DATA)
+    entry.add_to_hass(hass)
+
+    with _patch_media_player_device(mocked_device):
+        await hass.config_entries.async_setup(entry.entry_id)
+        await hass.async_block_till_done()
+
+    async def _assert_state():
+        state = hass.states.get(ENTITY_ID)
+        assert state.state == STATE_UNAVAILABLE
+
+    connect_change = MagicMock()
+    connect_change.exception = "disconnected"
+    type(mocked_device).get_supported_methods = AsyncMock(
+        side_effect=[SongpalException(""), SongpalException(""), _assert_state]
+    )
+    notification_callbacks = mocked_device.notification_callbacks
+    with patch("homeassistant.components.songpal.media_player.INITIAL_RETRY_DELAY", 0):
+        await notification_callbacks[ConnectChange](connect_change)
+    warning_records = [x for x in caplog.records if x.levelno == logging.WARNING]
+    assert len(warning_records) == 2
+    assert warning_records[0].message.endswith("Got disconnected, trying to reconnect.")
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5518609)